# Analysis of the results

Here we analyse the utility results of the StarGANv2-VC. Three experiments were run with different seeds.

In [1]:
import os
import numpy as np

In [2]:
# define the dirs that will be included in this analysis
log_dirs = [
    "../../logs/softvc/privacy",
    "../../logs/softvc/utility",
]

In [3]:
# map the components to the keys of the files that will be loaded
components = {
    "whisper-small": 4,
    "whisper-large": 4,
    "ser-audeering-w2v": 2,
    "nisqa": 2,
    "asv-plda/ignorant/results": 3,
    "asv-plda/lazy-informed/results": 3,
}

In [4]:
# aggregate results by the keys of the target and source speakers
results = dict()

for component, key in components.items():
    for log_dir in log_dirs:
        file = "eer.txt" if "privacy" in log_dir else "all.txt"
        if not os.path.exists(os.path.join(log_dir, "eval", component)):
            continue
        path = os.path.join(log_dir, "eval", component, file)
        with open(path, "r") as f:
            next(f)  # skip header

            for line in f:
                # get the value for the given dataset and key to the results
                elements = line.split()
                dataset = elements[0]
                value = elements[key]
                if dataset.endswith(".txt"):
                    dataset = dataset[:-4]
                # add the dataset to the results if needed
                if component not in results:
                    results[component] = dict()
                if dataset not in results[component]:
                    results[component][dataset] = list()

                results[component][dataset].append(round(float(value), 2))

In [5]:
# average the results of the different components
avgs = dict()
for component in results:
    avgs[component] = dict()
    for dataset in results[component]:
        avgs[component][dataset] = sum(results[component][dataset]) / len(results[component][dataset])

In [6]:
# print the results in a markdown table
print("| |", end=" ")  # empty cell in the top left corner
# print the dataset names as column headers
for dataset in avgs[list(avgs.keys())[0]]:
    print(dataset, end=" | ")
print()  # new line

# add hyphens to separate the header from the table
print("|", end=" ")
for i in range(len(avgs[list(avgs.keys())[0]]) + 1):
    print("---", end=" | ")
print()  # new line

# print the averages of each dataset for each ASV config
for component in avgs:
    print("|", end=" ")
    print(component, end=" | ")
    for dataset in avgs[component]:
        print(round(avgs[component][dataset], 2), end=" | ")
    print()  # new line


| | cv-test_3utts | ls-test-clean | edacc-test | ravdess | 
| --- | --- | --- | --- | --- | 
| whisper-small | 0.35 | 0.07 | 0.52 | 0.15 | 
| whisper-large | 0.28 | 0.06 | 0.47 | 0.08 | 
| ser-audeering-w2v | 1.0 | 1.0 | 1.0 | 1.0 | 
| nisqa | 3.49 | 3.95 | 3.6 | 3.92 | 
| asv-plda/ignorant/results | 0.43 | 0.44 | 0.44 | 0.4 | 
| asv-plda/lazy-informed/results | 0.28 | 0.19 | 0.3 | 0.1 | 


### Utility & Privacy results

| | cv-test_3utts | ls-test-clean | edacc-test | ravdess | 
| --- | --- | --- | --- | --- | 
| whisper-small | 0.35 | 0.07 | 0.52 | 0.15 | 
| whisper-large | 0.28 | 0.06 | 0.47 | 0.08 | 
| ser-audeering-w2v | 1.0 | 1.0 | 1.0 | 1.0 | 
| nisqa | 3.49 | 3.95 | 3.6 | 3.92 | 
| asv-plda/ignorant/results | 0.43 | 0.44 | 0.44 | 0.4 | 
| asv-plda/lazy-informed/results | 0.28 | 0.19 | 0.3 | 0.1 | 